File: Binary Tree Zigzag Level Order Traversal.py
# Leetcode Problem 103: Binary Tree Zigzag Level Order Traversal
# PYTHON CODE
from collections import deque

class Solution:
    def zigzagLevelOrder(self, root):
        if root is None:
            return []

        result = []
        queue = deque([root])
        left_to_right = True

        while queue:
            level = []

            for _ in range(len(queue)):
                node = queue.popleft()
                level.append(node.val)

                if node.left:
                    queue.append(node.left)

                if node.right:
                    queue.append(node.right)

            if not left_to_right:
                level.reverse()

            result.append(level)
            left_to_right = not left_to_right

        return result

File: Binary Tree Zigzag Level Order Traversal.cpp
// Leetcode Problem 103: Binary Tree Zigzag Level Order Traversal
// C++ CODE
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {

        vector<vector<int>> result;

        if (root == nullptr) {
            return result;
        }

        queue<TreeNode*> q;
        q.push(root);

        bool leftToRight = true;

        while (!q.empty()) {

            int size = q.size();
            vector<int> level;

            for (int i = 0; i < size; i++) {

                TreeNode* node = q.front();
                q.pop();

                level.push_back(node->val);

                if (node->left != nullptr) {
                    q.push(node->left);
                }

                if (node->right != nullptr) {
                    q.push(node->right);
                }
            }

            // Reverse every alternate level
            if (!leftToRight) {
                reverse(level.begin(), level.end());
            }

            result.push_back(level);

            leftToRight = !leftToRight;
        }

        return result;
    }
};